scripts/pylib/twister/twisterlib.py

```diff
@@ -787,9 +787,9 @@ def device_is_available(self, instance):
 
         return None
 
-    def make_device_available(self, serial):
+    def make_device_available(self, serial, serial_pty):
```

Comment:
Thanks for the review. @hakehuang , because we need to know whether the serial_pty is as the same as the d.serial_pty, to check this is the correct serial_pty you are going to set it available.  When we run by serial_pty,  the "serial" here is a string like this "/dev/ptyX", the d.serial_pty is the path of script like "xx/xx/catch_log.py".  This not only cannot be used to ensure the device but also the /dev/ptyX will not be the same each time we open a serial for it. 
Of course we can add another member in DUT such as d.curr_pty to save the pty port when we opened it each round but I think this is a simplest way to do that. Thanks!

Reply:
```suggestion
    def make_device_available(self, serial):
        if d.serial == serial or d.serial_pty == serial:
```

Reply:
Excuse me, @hakehuang ,  the d.serial_pty is something like "soc/xtensa/intel_adsp/tools/cavstwist.sh" and the serial will be "/dev/pty12",  so this won't work.

Reply:
per my understanding, the serial and serial_pty do not work together, so any one of then matched means the device is ready, right?
either d.serial == "/dev/pty12" or d.serial_pty == "soc/xtensa/intel_adsp/tools/cavstwist.sh", why it does not work?

Reply:
Hi @hakehuang , because if several devices run together, we cannot distinguish the devices by only judging the "not serial_pty".  The purpose here is to determine whether it is the correct device to make it available. 

Let me show you an example which in our use case:  (a hardware map file)
```
- connected: true
  id: None
  platform: intel_adsp_cavs15
  product: None
  runner: intel_adsp
  serial_pty: "soc/xtensa/intel_adsp/tools/cavstool_client.py,-s,192.168.1.2,-l"
  extra_params:
    - --remote-host=192.165.1.2
    - --run-twister

- connected: true
  id: None
  platform: intel_adsp_cavs18
  product: None
  runner: intel_adsp
  serial_pty: "soc/xtensa/intel_adsp/tools/cavstool_client.py,-s,192.168.1.3,-l"
  extra_params:
    - --remote-host=192.165.1.3
    - --run-twister

- connected: true
  id: None
  platform: intel_adsp_cavs25
  product: None
  runner: intel_adsp
  serial_pty: "soc/xtensa/intel_adsp/tools/cavstool_client.py,-s,192.168.1.4,-l"
  extra_params:
    - --remote-host=192.165.1.4
    - --run-twister
``` 
So the serial_pty here can be also used to distinguish which setting the right device to be available.

Actually,  this is my another proposal previously, using the device to judge the device. It also works:
```
-    def make_device_available(self, serial):
+    def make_device_available(self, serial, instance):
+        device = instance.platform.name
         for d in self.suite.duts:
+            if d.platform != device:
+                continue
             if d.serial == serial or d.serial_pty:
                 d.available = 1
```
Hope this can answer your question, thanks :)

Reply:
I still confused, in your example
intel_adsp_cavs15 ->  serial_pty: "soc/xtensa/intel_adsp/tools/cavstool_client.py,-s,192.168.1.2,-l"
intel_adsp_cavs18 -> "soc/xtensa/intel_adsp/tools/cavstool_client.py,-s,192.168.1.3,-l"
intel_adsp_cavs25 -> "soc/xtensa/intel_adsp/tools/cavstool_client.py,-s,192.168.1.4,-l"
so the make_device_available can check whether 
for intel_adsp_cavs15 
```
d.serial == "soc/xtensa/intel_adsp/tools/cavstool_client.py,-s,192.168.1.2,-l" or d.serial_pty == "soc/xtensa/intel_adsp/tools/cavstool_client.py,-s,192.168.1.2,-l"
```
which shall be ture
but if the dut is intel_adsp_cavs18
```
d.serial == "soc/xtensa/intel_adsp/tools/cavstool_client.py,-s,192.168.1.2,-l" or d.serial_pty == "soc/xtensa/intel_adsp/tools/cavstool_client.py,-s,192.168.1.2,-l"
```
which shall be false

so you just need to pass to the make_device_available with serial or pty_serial like below
if serial:
    make_device_available(serial)
else:
  make_device_available(pty_serial)

Reply:
Sorry for not making myself clearly, if I apply your suggestion like this, it still doesn't work. 

```
@@ -908,7 +915,11 @@ class DeviceHandler(Handler):
                 outs, errs = ser_pty_process.communicate()
                 logger.debug("Process {} terminated outs: {} errs {}".format(serial_pty, outs, errs))

-            self.make_device_available(serial_device)
+            if serial:
+                self.make_device_available(serial_device)
+            else:
+                self.make_device_available(serial_pty)
+
             return

         ser.flush()
@@ -1003,8 +1014,10 @@ class DeviceHandler(Handler):
         if post_script:
             self.run_custom_script(post_script, 30)

-        self.make_device_available(serial_device)
-
+        if serial:
+            self.make_device_available(serial_device)
+        else:
+            self.make_device_available(serial_pty)
```

Because in make_device_avalible(), these three device (cavs15/18/25), they all have d.serial_pty == true, so , it will make all of them available, but we just want one to be set to available, otherwise it will broken testing in another two boards.

Reply:
do you add  below change?
```
 if d.serial == serial or d.serial_pty == serial:
```
why it is true for platfrom that the serial_pty is not matched?

Reply:
Hi, I did not add the line, add it and trying.

Reply:
Hi @hakehuang , according to your suggestion and make minor change it works well by:
```
@@ -790,7 +790,7 @@ class DeviceHandler(Handler):

     def make_device_available(self, serial):
         for d in self.suite.duts:
-            if d.serial == serial or d.serial_pty:
+            if d.serial == serial or d.serial_pty == serial:
                 d.available = 1

     @staticmethod

@@ -908,7 +915,10 @@ class DeviceHandler(Handler):
                 outs, errs = ser_pty_process.communicate()
                 logger.debug("Process {} terminated outs: {} errs {}".format(serial_pty, outs, errs))

-            self.make_device_available(serial_device)
+            if serial_pty:
+                self.make_device_available(serial_pty)
+            else:
+                self.make_device_available(serial_device)
             return

         ser.flush()
@@ -1003,8 +1013,10 @@ class DeviceHandler(Handler):
         if post_script:
             self.run_custom_script(post_script, 30)

-        self.make_device_available(serial_device)
-
+        if serial_pty:
+            self.make_device_available(serial_pty)
+        else:
+            self.make_device_available(serial_device)

 class QEMUHandler(Handler):
```

Reply:
some urgly updates to make code tidy, you just need below
`self.make_device_available(serial_device + serial_device)`

Reply:
Good simplified but it's pity that the serial_device will not be None when serial_pty has value.
```
self.make_device_available(serial_device + serial_pty) or self.make_device_available(serial_pty + serial_device)
```